File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch02/Python/P02-MatOp.py
# Checks identity (A B)_trans = B_trans A_trans for random arrays A and B
from random import *
from matutil import *

# main

n = int(input("n = "))

A = [[0]*(n+1) for i in range(n+1)]                           # define arrays
B = [[0]*(n+1) for i in range(n+1)]
C = [[0]*(n+1) for i in range(n+1)]
D = [[0]*(n+1) for i in range(n+1)]

for i in range(1,n+1):                 # array A: random sub-unitary elements
   for j in range(1,n+1): A[i][j] = random()
print("Array A:")
MatPrint(A,n,n)

for i in range(1,n+1):                 # array B: random sub-unitary elements
   for j in range(1,n+1): B[i][j] = random()
print("Array B:")
MatPrint(B,n,n)

MatProd(A,B,C,n,n,n)                                                    # A*B
MatTrans(C,n)                                                   # (A*B)_trans

MatTrans(A,n)                                                       # A_trans
MatTrans(B,n)                                                       # B_trans
MatProd(B,A,D,n,n,n)                                       # B_trans* A_trans

MatDiff(C,D,D,n,n)                          # (A*B)_trans - B_trans * A_trans
print("Norm ((A*B)_trans - B_trans * A_trans) = ",MatNorm(D,n,n))

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch02/C/P02-MatOp.cpp
// Checks identity (A B)_trans = B_trans A_trans for random matrices A and B
#include <stdio.h>
#include "memalloc.h"
#include "matutil.h"

int main()
{
   double **A, **B, **C, **D;
   int i, j, n;

   printf("n = "); scanf("%i",&n);

   A = Matrix(1,n,1,n); B = Matrix(1,n,1,n);              // allocates arrays
   C = Matrix(1,n,1,n); D = Matrix(1,n,1,n);

   for (i=1; i<=n; i++)               // array A: random sub-unitary elements
      for (j=1; j<=n; j++) A[i][j] = rand()/(RAND_MAX+1e0);
   printf("Array A:\n");
   MatPrint(A,n,n);

   for (i=1; i<=n; i++)               // array B: random sub-unitary elements
      for (j=1; j<=n; j++) B[i][j] = rand()/(RAND_MAX+1e0);
   printf("Array B:\n");
   MatPrint(B,n,n);

   MatProd(A,B,C,n,n,n);                                               // A*B
   MatTrans(C,n);                                              // (A*B)_trans

   MatTrans(A,n);                                                  // A_trans
   MatTrans(B,n);                                                  // B_trans
   MatProd(B,A,D,n,n,n);                                  // B_trans* A_trans

   MatDiff(C,D,D,n,n);                     // (A*B)_trans - B_trans * A_trans
   printf("Norm ((A*B)_trans - B_trans * A_trans) = %e\n",MatNorm(D,n,n));
}
